### PR TITLE
[CAS-190] Refactor with Chris's changes

### PR DESCRIPTION
**Pull request by charlieyou**
Changes from Chris's branch
* Rebase on top of dev + merge in later changes
* Remove dead / unnecessary code (namely the rebalance module)
* Anchor doesn't support the use of generics as parameters to instructions or as fields in account structs. This means that `BackendContainer` can't be an input param nor stored in the `Vault` struct. The solution implemented was to keep the existing structs `StrategyWeightsArg` and `Allocations` and add the associated `From` impls.
* Allocations are passed around in `rebalance` as `u64` instead of `SlotTrackedValue`
* BackendContainerGeneric::Default fills array with Some instead of None

### CODE REVIEW

File: programs/castle-vault/src/asset_container/rate.rs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+            .map(|(_, r)| r)
+            .max()
+            .ok_or(ErrorCode::InvalidProposedWeights)?;
+
+        let sum = self
+            .into_iter()
+            .map(|(_, r)| r)
+            .try_fold(Rate::zero(), |acc, x| acc.try_add(*x))?;
+
+        (sum == Rate::one() && max <= cap).as_result((), ErrorCode::InvalidProposedWeights.into())
```
Comment (charlieyou) [inline programs/castle-vault/src/asset_container/rate.rs:25]: TODO replace with ok_or

Comment (charlieyou) [inline programs/castle-vault/src/asset_container/rate.rs:25]: TODO replace with ok_or

File: programs/castle-vault/Cargo.toml [hunk 2]
```diff
@@ -32,7 +32,9 @@ no-idl = []
 [dependencies]
 anchor-lang = "0.18.2"
 anchor-spl = "0.18.2"
+arrayref = "0.3"
```
Comment (nascent-chris) [inline programs/castle-vault/Cargo.toml:35]: Ended up not using this, forgot to take it out -- sorry about that

This is a handy crate to keep in mind, though.

File: programs/castle-vault/src/asset_container/iter.rs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        AssetContainerIterator {
+            inner: self,
+            inner_iter: Provider::iter(),
+        }
+    }
+}
+
+impl<T, const N: usize> IntoIterator for AssetContainerGeneric<T, N> {
+    type Item = (Provider, T);
+    type IntoIter = OwnedBackendContainerIterator<T, N>;
```
Comment (nascent-chris) [inline programs/castle-vault/src/asset_container/iter.rs:23]:
```suggestion
    type IntoIter = OwnedAssetContainerIterator<T, N>;
```

File: programs/castle-vault/src/asset_container/iter.rs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    type Item = (Provider, &'inner T);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.inner_iter
+            .next()
+            .map(|provider| (provider, &self.inner[provider]))
+    }
+}
+
+pub struct OwnedBackendContainerIterator<T, const N: usize> {
```
Comment (nascent-chris) [inline programs/castle-vault/src/asset_container/iter.rs:48]:
```suggestion
pub struct OwnedAssetContainerIterator<T, const N: usize> {
```

File: programs/castle-vault/src/asset_container/iter.rs [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    inner: AssetContainerGeneric<T, N>,
+    inner_iter: ProviderIter,
+}
+
+impl<T, const N: usize> Iterator for OwnedBackendContainerIterator<T, N> {
```
Comment (nascent-chris) [inline programs/castle-vault/src/asset_container/iter.rs:53]:
```suggestion
impl<T, const N: usize> Iterator for OwnedAssetContainerIterator<T, N> {
```

File: programs/castle-vault/src/asset_container/iter.rs [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    type Item = (Provider, T);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.inner_iter.next().map(|provider| {
+            (
+                provider,
+                self.inner.inner[provider as usize]
+                    .take()
+                    .expect("missing index in OwnedBackendContainerIterator"),
```
Comment (nascent-chris) [inline programs/castle-vault/src/asset_container/iter.rs:62]:
```suggestion
                    .expect("missing index in OwnedAssetContainerIterator"),
```

File: programs/castle-vault/src/asset_container/mod.rs [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    pub(crate) inner: [Option<T>; N],
+}
+
+impl<T, const N: usize> AssetContainerGeneric<T, N> {
+    pub fn len(&self) -> usize {
+        N
+    }
+
+    pub fn is_empty(&self) -> bool {
+        // TODO: Should this just always return `true`, or do we want it to mean "uninitialized"?
```
Comment (nascent-chris) [inline programs/castle-vault/src/asset_container/mod.rs:31]: May be worth adding a comment communicating that `is_empty()` means "uninitialized"

File: programs/castle-vault/src/asset_container/mod.rs [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+    }
+}
+
+impl<T, const N: usize> Index<Provider> for AssetContainerGeneric<T, N> {
+    type Output = T;
+
+    fn index(&self, index: Provider) -> &Self::Output {
+        self.inner[index as usize].as_ref().unwrap_or_else(|| {
+            panic!(
+                "missing index {:?} / {:?} in BackendContainerGeneric",
```
Comment (nascent-chris) [inline programs/castle-vault/src/asset_container/mod.rs:42]:
```suggestion
                "missing index {:?} / {:?} in AssetContainerGeneric",
```

File: programs/castle-vault/src/asset_container/mod.rs [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+            )
+        })
+    }
+}
+
+impl<T, const N: usize> IndexMut<Provider> for AssetContainerGeneric<T, N> {
+    fn index_mut(&mut self, index: Provider) -> &mut Self::Output {
+        self.inner[index as usize].as_mut().unwrap_or_else(|| {
+            panic!(
+                "missing index {:?} / {:?} in BackendContainerGeneric",
```
Comment (nascent-chris) [inline programs/castle-vault/src/asset_container/mod.rs:53]:
```suggestion
                "missing index {:?} / {:?} in AssetContainerGeneric",
```

File: programs/castle-vault/src/instructions/init.rs [hunk 10]
```diff
@@ -194,7 +194,7 @@ pub fn handler(
     ctx.accounts.validate_referral_token()?;
 
     let vault = &mut ctx.accounts.vault;
-    vault.version = [2, 0, 0];
+    vault.version = [2, 0, 1];
```
Comment (nascent-chris) [inline programs/castle-vault/src/instructions/init.rs:197]: Unrelated to the PR, but may be helpful

```suggestion
    vault.version = [
        env!("CARGO_PKG_VERSION_MAJOR")
            .parse::<u8>()
            .expect("failed to parse major version"),
        env!("CARGO_PKG_VERSION_MINOR")
            .parse::<u8>()
            .expect("failed to parse minor version"),
        env!("CARGO_PKG_VERSION_PATCH")
            .parse::<u8>()
            .expect("failed to parse patch version"),
    ];
```

**Review by nascent-chris — APPROVED** [review]
Nicely cleaned up!  Also `AssetContainer` is a way better name.

No real comments (other than some missed renaming).  I also added an example of using the `env!()` to build the version array in `init::handler()`.

Thanks for allowing me to contribute!  Glad to see it being useful

**Pull request merged**